src/main/scala/is/hail/expr/ir/functions/IntervalFunctions.scala

```diff
@@ -65,14 +64,118 @@ object IntervalFunctions extends RegistryFunctions {
+        val interval = new IRInterval(mb, TInterval(pointType), intTriplet.value[Long])
+        val compare = interval.ordering(CodeOrdering.compare)
+
+        val contains = Code(
+          interval.storeToLocal,
+          mPoint := pointTriplet.m,
+          vPoint.storeAny(pointTriplet.v),
+          cmp := compare((mPoint, vPoint), interval.start),
+          (cmp > 0 || (cmp.ceq(0) && interval.includeStart)) && Code(
+            cmp := compare((mPoint, vPoint), interval.end),
+            cmp < 0 || (cmp.ceq(0) && interval.includeEnd)))
+
+        EmitTriplet(
+          Code(intTriplet.setup, pointTriplet.setup),
+          intTriplet.m,
+          contains)
+    }
+
+    registerCode("isEmpty", TInterval(tv("T")), TBoolean()) {
+      case (mb, intOff) =>
+        val interval = new IRInterval(mb, TInterval(tv("T").t), intOff)
+
+        Code(
+          interval.storeToLocal,
+          interval.isEmpty
+        )
+    }
+
+    registerCode("overlaps", TInterval(tv("T")), TInterval(tv("T")), TBoolean()) {
+      case (mb, iOff1, iOff2) =>
+        val pointType = tv("T").t
+
+        val interval1 = new IRInterval(mb, TInterval(pointType), iOff1)
+        val interval2 = new IRInterval(mb, TInterval(pointType), iOff2)
+
+        Code(
+          interval1.storeToLocal,
+          interval2.storeToLocal,
+          !(interval1.isEmpty || interval2.isEmpty ||
+            interval1.isBelow(interval2, checkEmpty = false) ||
+            interval1.isAbove(interval2, checkEmpty = false))
+        )
+    }
+  }
+}
+
+class IRInterval(mb: EmitMethodBuilder, typ: TInterval, value: Code[Long]) {
+  val ref: LocalRef[Long] = mb.newLocal[Long]
+  val region: Code[Region] = IntervalFunctions.getRegion(mb)
+
+  def ordering[T](op: CodeOrdering.Op): ((Code[Boolean], Code[_]), (Code[Boolean], Code[_])) => Code[T] =
+    mb.getCodeOrdering[T](typ.pointType, op, missingGreatest = true)(region, _, region, _)
+
+  def storeToLocal: Code[Unit] = ref := value
+
+  def start: (Code[Boolean], Code[_]) =
+    (!typ.startDefined(region, ref), region.getIRIntermediate(typ.pointType)(typ.startOffset(ref)))
+  def end: (Code[Boolean], Code[_]) =
+    (!typ.endDefined(region, ref), region.getIRIntermediate(typ.pointType)(typ.endOffset(ref)))
+  def includeStart: Code[Boolean] = typ.includeStart(region, ref)
+  def includeEnd: Code[Boolean] = typ.includeEnd(region, ref)
+
+  def isEmpty: Code[Boolean] = {
+    val gt = ordering(CodeOrdering.gt)
+    val gteq = ordering(CodeOrdering.gteq)
+
+    (includeStart && includeEnd).mux(
+      gt(start, end),
+      gteq(start, end))
+  }
+
+  def isAbove(other: IRInterval, checkEmpty: Boolean = true): Code[Boolean] = {
+    val cmp = mb.newLocal[Int]
+    val compare = ordering(CodeOrdering.compare)
+    val above = Code(
+      cmp := compare(start, other.end),
+      cmp > 0 || (cmp.ceq(0) && (!includeStart || !other.includeEnd)))
+
+    if (checkEmpty)
+      this.isEmpty || other.isEmpty || above
```

Comment:
is there going to be a use case for `isAbove` where `checkEmpty` will be true? what's the point of checking this here?

Reply:
I can't think of one off the top of my head, but this is consistent with the definition on `Interval`. I just didn't want to duplicate the `isEmpty` logic where unnecessary. I could remove this flag and rename the methods to `isAboveOnNonemptyIntervals` if that would be better.

Reply:
I have a slight preference for the second, since that parameter will likely not be used.